addons/controllers/calico/calicoconfig_utils.go

```diff
@@ -33,6 +40,56 @@ type config struct {
 	SkipCNIBinaries bool   `yaml:"skipCNIBinaries"`
 }
 
+// ClusterToCalicoConfig returns a list of Requests with CalicoConfig ObjectKey
+func (r *CalicoConfigReconciler) ClusterToCalicoConfig(o client.Object) []ctrl.Request {
+	cluster, ok := o.(*clusterv1beta1.Cluster)
+	if !ok {
+		r.Log.Error(errors.New("invalid type"),
+			"Expected to receive Cluster resource",
+			"actualType", fmt.Sprintf("%T", o))
+		return nil
+	}
+
+	r.Log.V(4).Info("Mapping cluster to CalicoConfig")
+
+	configs := &cniv1alpha1.CalicoConfigList{}
+
+	if err := r.Client.List(context.Background(), configs); err != nil {
+		r.Log.Error(err, "Error listing CalicoConfig")
+		return nil
+	}
+
+	var requests []ctrl.Request
+	for i := range configs.Items {
+		config := &configs.Items[i]
+		if config.Namespace == cluster.Namespace {
+			// avoid enqueuing reconcile requests for template CalicoConfig CRs in event handler of Cluster CR
+			if _, ok := config.Annotations[constants.TKGAnnotationTemplateConfig]; ok && config.Namespace == r.Config.SystemNamespace {
+				continue
+			}
+
+			// corresponding CalicoConfig should have following ownerRef
+			ownerReference := metav1.OwnerReference{
+				APIVersion: clusterv1beta1.GroupVersion.String(),
+				Kind:       cluster.Kind,
+				Name:       cluster.Name,
+				UID:        cluster.UID,
+			}
+
+			if clusterapiutil.HasOwnerRef(config.OwnerReferences, ownerReference) || config.Name == fmt.Sprintf("%s-%s-package", cluster.Name, constants.CalicoAddonName) {
```

Comment:
https://github.com/vmware-tanzu/tanzu-framework/pull/3318#discussion_r976341528

Link the related comment for cpi controller here